Analyzer and metric collector visit the repository together - #25

jan-gerling opened the pull request:
*  refactored App.run visit the repositories only once and collect all necessary data together
 *  removed addToList from ProcessMetricsCollector
 *  refactored collect from ProcessMetricsCollector into collectMetrics(commit, yesSet, isRefactoring)
 *  made collectCommitData from RefactoringAnalyzer return the yesList

mauricioaniche commented on data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java:
```diff
@@ -51,73 +47,45 @@ public ProcessMetricsCollector(Project project, Database db, Repository reposito
 		this.branch = branch;
 		this.fileStoragePath = FilePathUtils.lastSlashDir(fileStoragePath);
 		this.lastCommitToProcess = lastCommitToProcess;
-		todo = new HashMap<>();
```
One huge in-memory list less... Beautiful!!!

jan-gerling commented on data-collection/src/main/java/refactoringml/RefactoringAnalyzer.java:
```diff
@@ -53,10 +50,11 @@ public RefactoringAnalyzer (Project project, Database db, Repository repository,
 
 		if (commit.getParentCount() == 0 || !studied(refactoring)) {
-			return ;
+			//TODO: check if this is correct and desired behavior
```
I find it a bit difficult to test, because I am not sure which behavior I would expect in this case.

jan-gerling commented on data-collection/src/main/java/refactoringml/App.java:
```diff
@@ -136,21 +138,24 @@ public Project run () throws Exception {
 						try {
 							db.openSession();
-							refactoringAnalyzer.collectCommitData(currentCommit, ref);
+							allYeses = refactoringAnalyzer.collectCommitData(currentCommit, ref);
```
> This possible bug makes me think that, before merging this PR, we should write a test where two refactorings have happened in the same commit!

Done, and you were right.

jan-gerling replied:
Fixed this.

jan-gerling commented:
> Wow, amazing that you tackled this one! It's a tough one!!
> 
> I think I found a small bug there! See my comment below!

I think I fixed the bug and added a test for this case.

mauricioaniche commented on data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java:
```diff
@@ -51,73 +47,44 @@ public ProcessMetricsCollector(Project project, Database db, Repository reposito
-				log.debug("Commit ID " + commit.getName());
-				RevCommit commitParent = commit.getParentCount() == 0 ? null : commit.getParent(0);
+	public void collectMetrics(RevCommit commit, Set<Long> allYeses, boolean isRefactoring) throws IOException {
+		if(commit.getParentCount() > 1) return;
```
Just a note: maybe we should skip merge commits in the "main walker" in the App file. What do you think?

mauricioaniche replied:
And if we do that there, then there's no need to repeat it here

mauricioaniche commented on data-collection/src/main/java/refactoringml/ProcessMetricsCollector.java:
```diff
@@ -51,73 +47,44 @@ public ProcessMetricsCollector(Project project, Database db, Repository reposito
-				} finally {
-					db.close();
-				}
+		// if the class happened to be refactored, then, print its process metrics at that time
```
maybe the comment here should be about "if this commit contained a refactoring, then collect its process metrics"
